node/rpc/src/mpool_api.rs

```diff
@@ -0,0 +1,176 @@
+    let address = Address::from_str(&addr_str)?;
+    let sequence = data.mpool.get_sequence(&address).await?;
+    Ok(sequence)
+}
+
+/// Return Vec of pending messages in mpool
+pub(crate) async fn pending<DB, KS, MP>(
+    data: Data<RpcState<DB, KS, MP>>,
+    Params(params): Params<(TipsetKeys,)>,
+) -> Result<Pending, JsonRpcError>
+where
+    DB: BlockStore + Send + Sync + 'static,
+    KS: KeyStore + Send + Sync + 'static,
+    MP: Provider + Send + Sync + 'static,
+{
+    let (tsk,) = params;
+    let mut ts = chain::tipset_from_keys(data.store.as_ref(), &tsk)?;
+
+    let (mut pending, mpts) = data.mpool.pending().await?;
+
+    let mut have_cids = HashSet::new();
+    for item in pending.iter() {
+        have_cids.insert(item.cid()?);
+    }
+
+    if mpts.epoch() > ts.epoch() {
+        return Ok(Pending { msgs: pending });
+    }
+
+    loop {
+        if mpts.epoch() == ts.epoch() {
+            if mpts == ts {
+                return Ok(Pending { msgs: pending });
+            }
+
+            // mpts has different blocks than ts
+            let have = data.mpool.as_ref().messages_for_blocks(ts.blocks()).await?;
+
+            for sm in have {
+                have_cids.insert(sm.cid()?);
+            }
+        }
+
+        let msgs = data.mpool.as_ref().messages_for_blocks(ts.blocks()).await?;
+
+        for m in msgs {
+            if have_cids.contains(&m.cid()?) {
+                continue;
+            }
+
+            have_cids.insert(m.cid()?);
+            pending.push(m);
+        }
+
+        if mpts.epoch() >= ts.epoch() {
+            return Ok(Pending { msgs: pending });
+        }
+
+        let mut headers: Vec<BlockHeader> = Vec::new();
+        let parents = ts.parents().cids();
+        for cid in parents {
+            let block: BlockHeader = data
+                .store
+                .as_ref()
+                .get(cid)?
+                .ok_or("can't find block with cid")?;
+            headers.push(block);
+        }
+        ts = Tipset::new(headers)?;
+    }
+}
+
+/// Add SignedMessage to mpool, return msg CID
+pub(crate) async fn push<DB, KS, MP>(
+    data: Data<RpcState<DB, KS, MP>>,
+    Params(params): Params<(SignedMessageJson,)>,
+) -> Result<CidJson, JsonRpcError>
+where
+    DB: BlockStore + Send + Sync + 'static,
+    KS: KeyStore + Send + Sync + 'static,
+    MP: Provider + Send + Sync + 'static,
+{
+    let (SignedMessageJson(smsg),) = params;
+
+    let cid = data.mpool.as_ref().push(smsg).await?;
+
+    Ok(CidJson(cid))
+}
+
+/// Sign given UnsignedMessage and add it ot mpool, return SignedMessage
+pub(crate) async fn push_message<DB, KS, MP>(
+    data: Data<RpcState<DB, KS, MP>>,
+    Params(params): Params<(UnsignedMessageJson,)>,
+) -> Result<SignedMessageJson, JsonRpcError>
+where
+    DB: BlockStore + Send + Sync + 'static,
+    KS: KeyStore + Send + Sync + 'static,
+    MP: Provider + Send + Sync + 'static,
+{
+    let (UnsignedMessageJson(umsg),) = params;
+
```

Comment:
can you add a TODO somewhere in this function for handling defaults for sequence, gas limit, and gas price? Seems the inconsistent one is that Lotus disallows non 0 nonce messages for this method, probably should match for consistent behaviour?